korok/korok.py

```diff
@@ -51,7 +81,89 @@ def fit(
             metric=Metric.COSINE,
             **kwargs,
         )
-        return cls(encoder=encoder, vicinity=vicinity, reranker=reranker)
+
+        # Add bm25s if alpha > 0
+        if cls.is_hybrid_available() and alpha < 1.0:
+            global bm25s
+            import bm25s
+
+            bm25 = bm25s.BM25()
+            tokens = bm25s.tokenize(texts, stopwords=stopwords)
+            bm25.index(tokens)
+        else:
+            bm25 = None
+
+        return cls(encoder=encoder, vicinity=vicinity, reranker=reranker, bm25=bm25, alpha=alpha, corpus=texts)
+
+    def _normalize_scores(self, scores: np.ndarray) -> np.ndarray:
+        """Normalize the scores to be between 0 and 1."""
+        # Get the min and max scores for each document
+        min_score = np.min(scores, axis=-1).reshape(-1, 1)
+        max_score = np.max(scores, axis=-1).reshape(-1, 1)
+
+        # Normalize the scores
+        denom = max_score - min_score
+        numerator = scores - min_score
+
+        return numerator / denom
+
+    def _split_vicinity_results(self, results: List[List[Tuple[str, float]]]) -> List[List[Tuple[str, float]]]:
+        """Split the results from vector search into two lists."""
+        vicinity_docs = []
+        vicinity_scores = []
+        for result in results:
+            for doc, score in result:
+                vicinity_docs.append(doc)
+                vicinity_scores.append(score)
+        vicinity_scores = np.array(vicinity_scores)
+        return vicinity_docs, vicinity_scores
+
+    def _merge_results(
+        self,
+        vicinity_results: List[List[Tuple[str, np.float64]]],
+        bm25_results: Tuple[List[List[str]], np.ndarray],
+        k: int = 10,
+    ) -> List[List[tuple[str, float]]]:
+        """Merge the results from vector search and bm25 search."""
+        # Initialize the scores list
+        scores_list = [{} for i in range(len(vicinity_results))]
```

Comment:
You now have the document class, so you could turn this into a `defaultdict` with `Document` as a default. That way, you don't need to check below if a certain key exists.

Reply:
Looking at it again, I don't think there's a need to pre-populate this list, you can just append to it as you go along.

Reply:
Yes! That's a great suggestion; I didn't think of that earlier~ Since we don't have to check for the document if it exists or not anymore, it saves a lot of time...